### ISSUE TITLE
Bus fault when importing ONNX model on Android

### ISSUE THREAD

**Issue by Palmitoxico** [opencv/opencv#16373]
##### System information (version)
- OpenCV => 1a74de1f9c7a3e1c8409a8b8e20c1f992f614e4a (4.2)
- Operating System / Platform => Android 4.4.4 amv7l
- Compiler => Android NDK r18b (clang, Android API level = 16)
- Optmization => -DCMAKE_BUILD_TYPE=Release (-Oz)

##### Detailed description

I'm getting a SIGBUS when loading an ONNX model on Android (armv7l), this problem only occurs when building with optimizations enabled, -O0 doesn't trigger the issue. I've nailed down the problem to these specific lines:

https://github.com/opencv/opencv/blob/d85e67d3ecf425f0ffbf643e7fa8323e289bfff0/modules/dnn/src/onnx/onnx_importer.cpp#L149-L153

The problem is that the pointer returned by std::string::c_str() doesn't have any guarantee that it is 8 bytes aligned, and sure enough when inspecting with GDB the ```val``` pointer isn't aligned, thus ```src```.

I've workaround this by allocating a intermediate buffer and copying the data pointed by ```val```:

```c++
{
    char* val = const_cast<char*>(tensor_proto.raw_data().c_str());
    int size = tensor_proto.raw_data().size();
    int64_t* src = new int64_t[size];
    memcpy(src, val, size);
    convertInt64ToInt32(src, dst, blob.total());
    delete[] src;
}
```

I don't known if this is an good solution, please let me known.

##### Steps to reproduce

Build for Android armv7l, NDK r18b with optimizations. Call ```cv::dnn::readNetFromONNX()``` with a valid ONNX file.

Thanks,
Augusto.

**Comment by alalek** [comment]
Right, ARMv7 doesn't tolerate unaligned memory access (unchecked `char*` => `int64_t*`).

**Comment by alalek** [comment]
@Palmitoxico Thank you for the report and investigation!
Could you please try this patch: #16377 ?

**Comment by Palmitoxico** [comment]
@alalek The #16377 patch doesn't solve the issue, still getting SIGBUS when deferencing the src pointer. Are the ```unaligned_int64_t``` type supposed to force the compiler to access the data in a byte-per-byte basis?

**Comment by alalek** [comment]
Thank you! This doesn't work because templates can't handle attributes properly.
Please validate updated patch.

**Comment by Palmitoxico** [comment]
@alalek #16377 works now, thanks!

**Comment by amichayfeld** [comment]
I got error when isAligned function is called (it is not defined).
Where can i find it?
Meanwhile I wrote one, but I did not figure out how you can run this code..

**Comment by amichayfeld** [comment]
Hi. This issue occurs also in version 4.1.0
